dre-service/src/main/java/io/cdap/re/RulesDB.java

```diff
@@ -147,376 +199,399 @@ public void updateRule(String id, RuleRequest rule) throws RuleNotFoundException
       );
     }
 
-    if (rule.getThen() == null || rule.getWhen().trim().isEmpty()) {
+    if (rule.getThen().trim().isEmpty()) {
       throw new IllegalArgumentException(
         String.format("Rule '%s' requires the mandatory field 'then'", rule.getId())
       );
     }
 
-    byte[][] columns = new byte[][] {
-      DESCRIPTION, CONDITION, ACTION, UPDATED
-    };
+    Collection<Field<?>> updatedRuleFields = new ArrayList<>();
+    updatedRuleFields.add(Fields.stringField(ID_COL, id));
+    updatedRuleFields.add(Fields.stringField(DESCRIPTION_COL, rule.getDescription()));
+    updatedRuleFields.add(Fields.stringField(CONDITION_COL, rule.getWhen()));
+    updatedRuleFields.add(Fields.stringField(ACTION_COL, rule.getThen()));
+    updatedRuleFields.add(Fields.longField(UPDATED_COL, getCurrentTime()));
 
-    byte[][] values = new byte[][] {
-      Bytes.toBytes(rule.getDescription()),
-      Bytes.toBytes(rule.getWhen()),
-      Bytes.toBytes(rule.getThen()),
-      Bytes.toBytes(getCurrentTime())
-    };
-
-    rules.put(toKey(id), columns, values);
+    rulesTable.upsert(updatedRuleFields);
   }
 
-  public Map<String, Object> retrieveRule(String id) throws RuleNotFoundException {
-    Map<String, Object> result = new TreeMap<>();
-    Row row = rules.get(toKey(id));
-    if (row.isEmpty()) {
+  Map<String, Object> retrieveRule(String id) throws RuleNotFoundException, IOException {
```

Comment:
Nothing to do with your PR, but this is a poorly designed API, it should be returning a class that represents a Rule, not some Map of objects.

Is this called in many places? If it's not much work it probably makes sense to rename RuleRequest to Rule and make this method return a Rule. Doesn't have to be in this PR though.

Reply:
It's called in a couple of places, but I will make that change in a subsequent PR.